src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -501,29 +504,23 @@ repl_req_ptr_t RaftReplDev::applier_create_req(repl_key const& rkey, journal_typ
     // We need to allocate the block, since entry doesn't exist or if it exist, two threads are trying to do the same
     // thing. So take state mutex and allocate the blk
     std::unique_lock< std::mutex > lg(rreq->m_state_mtx);
-    rreq->init(rkey, code, false /* is_proposer */, user_header, key, data_size);
-
-    // There is no data portion, so there is not need to allocate
+    auto status = rreq->init(rkey, code, false /* is_proposer */, user_header, key, data_size, m_listener);
     if (!rreq->has_linked_data()) { return rreq; }
-    if (rreq->has_state(repl_req_state_t::BLK_ALLOCATED)) { return rreq; }
```

Comment:
this line is trying to skip the allocation if it is already allocated, for example , data channel comes first, the `happened == true`, however raft channel comes in with `happened = false` but  finished the allocation, then data channel should skip the allocation.

similar logic is needed for rreq->init as you move the allocation into that.

Reply:
Yes, should add this logic in rreq->init.